### PR TITLE
Start test suite.

### PR DESCRIPTION
**Pull request by mkcor**
Addressing #5, we want this package (collection of utility functions for `dash-bio`) covered with unit tests.

Dear @byronz and @shammamah,

Before proceeding further, I would love your feedback on the following:

* structure and naming convention:
  - one test file per module (`gene_expression_reader.py` --> `test_gene_expression_reader.py`);
  - test functions named after function and specific feature (`read_soft(datapath_or_datastring, is_datafile=True, return_filtered_data=False, rows=None, columns=None)` --> `test_read_soft_datastring()`, `test_read_soft_filtered()`, and more to come).

* sample data for testing purposes:
  I have included a string of sample SOFT data as a global variable, but I'm considering reading files over the Internet as well, once Dash Bio datasets are up on https://github.com/plotly/datasets;

* conda environment:
  I have a pro-conda bias (inherited from the SciPy/PyData community) so, since we had no explicit setup instructions, I decided to run the embryonic test suite in a conda env. That's what I documented.

### CODE REVIEW

File: tests/test_gene_expression_reader.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+'10882\n1007_s_at\tMIR4640\t24.8302\t25.4542\t25.6869\t25.7307\n1053_at' + \
+'\tRFC2\t25.6556\t25.7486\t25.9348\t25.8511\n117_at\tHSPA6\t22.3986\t21' + \
+'.4459\t22.136\t21.6938\n121_at\tPAX8\t24.5789\t24.5075\t24.5385\t24.57' + \
+'36\n1255_g_at\tGUCA1A\t21.416\t21.0745\t21.0107\t21.1167\n1294_at\tMIR' + \
+'5193\t23.6715\t23.4425\t22.9409\t22.9591\n1316_at\tTHRA\t22.7604\t23.1' + \
+'194\t23.0321\t23.0469\n'
+
+
+def test_read_soft_datastring():
+    """Test that SOFT data string can be read."""
```
Comment (byronz) [inline tests/test_gene_expression_reader.py]: if you need a doc string for unit test, probably consider making the test function name more clear.  like `test_soft_datastring_is_readable` ?

File: tests/test_gene_expression_reader.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from __future__ import absolute_import
+
+from dash_bio_utils.gene_expression_reader import read_soft
+
+# This piece of toy data corresponds to the first 2161 characters of data file:
+# url = 'https://raw.githubusercontent.com/plotly/dash-bio/master/tests/dashbio_demos/sample_data/'
+# url = url + 'clustergram_GDS5373.soft'
+DATASTRING = '^DATABASE = Geo\n!Database_name = Gene Expression Omnibus' + \
```
Comment (byronz) [inline tests/test_gene_expression_reader.py]: minor, this can be put as an asset file or return as a fixture in conftest.py if you need that across multiple test files.

File: README.md [hunk 3]
```diff
@@ -17,4 +17,15 @@ and more, please [reach out](https://plotly.typeform.com/to/mH1Cpb).
...
 If you would like to contribute to this repository, please refer to
 the [contributing
-guidelines](https://github.com/plotly/dash-bio-utils/blob/master/CONTRIBUTING.md).
\ No newline at end of file
+guidelines](https://github.com/plotly/dash-bio-utils/blob/master/CONTRIBUTING.md).
+
+### Running tests
+
+```bash
+conda create -n dash-bio-utils python=3.7
```
Comment (shammamah) [inline README.md]: Other Dash repos use `virtualenv`. I think it might be better to keep consistent with that.

File: tests/test_gene_expression_reader.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    description = soft[0]
+    assert 'title' in description.keys()
+
+    subsets = soft[1]
+    assert isinstance(subsets, dict)
+
+    rows = soft[2]
+    # first row is '1007_s_at'
+    assert '1007_s_at' in rows
```
Comment (byronz) [inline tests/test_gene_expression_reader.py]: if the criterion is  `1007_s_at is the first row`, then the following assert is not checking exactly that.

File: tests/test_gene_expression_reader.py (hunk 2 above)
Comment (shammamah) [reply tests/test_gene_expression_reader.py]: ^ Agreed -- it might be better to put it in a separate file.

File: tests/test_gene_expression_reader.py (hunk 4 above)
Comment (shammamah) [inline tests/test_gene_expression_reader.py]: We might want to check the number of rows as well.

File: tests/test_gene_expression_reader.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+def test_read_soft_filtered():
+    """Test that SOFT data string can be filtered."""
+    soft = read_soft(DATASTRING,
+                     is_datafile=False,
+                     return_filtered_data=True,
+                     rows=['1007_s_at', '1053_at', '117_at'],
+                     columns=['GSM1110880', 'GSM1110881'])
+
+    # when filtering, return only the filtered data
+    assert soft.shape[0] == 3
```
Comment (shammamah) [inline tests/test_gene_expression_reader.py]: We could try checking whether the values are correct, too!

File: tests/test_gene_expression_reader.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    # first row is '1007_s_at'
+    assert '1007_s_at' in rows
+
+    cols = soft[3]
+    # SOFT reader keeps columns with label containing 'GSM'
+    assert all(['GSM' in col for col in cols])
+
+
+def test_read_soft_filtered():
+    """Test that SOFT data string can be filtered."""
```
Comment (byronz) [inline tests/test_gene_expression_reader.py]: same here, docstring can be 🔪  with a longer test function name

Comment (byronz) [reply tests/test_gene_expression_reader.py]: another tip for the unitary test function name,  if we end up making a very long test name, it's a good sign we should probably split the scope.

**Comment by byronz** [comment]
@mkcor  do you plan to add tests handling for example data with bad format or other edge/boundary cases?

File: README.md (hunk 3 above)
Comment (byronz) [reply README.md]: virtualenv is a comproise for python 2.7,   py3 venv is the default way to go

Comment (shammamah) [reply README.md]: @byronz Interesting! I didn't know that :P

**Comment by byronz** [comment]
> > test functions named after function and specific feature
> 
> We can name them according to the convention we have specified in the docs: https://dash.plot.ly/testing (under number 2 in "Notes").
> 

That test function name convention is more for integration tests.  for unit test, you can still follow the pattern except the `tcid`, I feel it's an overkill to give each unit test a `tcid`.  

-  if you are doing exactly the [pyramid model](https://martinfowler.com/articles/practical-test-pyramid.html), the number of tests would be significantly large.
-  some of the pros from tcid is not true in unit test context, they are much faster and more robust if you don't change the API.

File: README.md (hunk 3 above)
Comment (mkcor) [reply README.md]: I'm aware of it, but questions about using conda come up every now and then (reflecting the community that finds Dash and that Dash finds, reciprocally), so I thought it could be a nice, soft signal... since `dash-bio-utils` is not (meant to be) highly visible in the Dash ecosystem, and I haven't written setup/install instructions, only instructions for running tests.

Personally, I like being flexible and keeping up-to-date so I'm happy to run both virtualenvs and conda envs, depending. But I can convert these commands into a virtualenv + pip version if you guys prefer.

File: tests/test_gene_expression_reader.py (hunk 2 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: If it lives in a file, we'll have to do
```py
string = open(<filepath>, 'r').read()
read_soft(string, is_datafile=False)
```
to test passing data as a raw string, and we'll depend on the data source, but why not.

> the main concern here is the split of the long string could be error-prone.

Sure; in this case, I have done it manually, so it works by design, but that doesn't scale... at the same time, should it scale? For now, my approach is precisely to craft each test manually and look at them with my own eyes, edit them as needed with my own hands... What do you guys think?

> return as a fixture in conftest.py if you need that across multiple test files

If we keep this file structure (which is why I asked for your feedback), we shouldn't need it across multiple test files, because all SOFT-related tests would live in this same file. Nonetheless, I would love to learn something new and use a fixture in conftest.py: @byronz could you please point me to a starting point or some docs? Thanks!

Comment (mkcor) [reply tests/test_gene_expression_reader.py]: Meanwhile, I have submitted https://github.com/plotly/datasets/pull/24 to have the data file available somewhere (`DATASTRING` corresponds to the first 2161 characters of this data file).

But I'm wondering if we couldn't read the data directly from https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS5373 -- @shammamah do you know if they have an API?

File: tests/test_gene_expression_reader.py (hunk 4 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: True; fixed in 552f52e44a.

Comment (mkcor) [reply tests/test_gene_expression_reader.py]: Done in 552f52e44a.

File: README.md (hunk 3 above)
Comment (mkcor) [reply README.md]: Note that the Dash docs are 'bilingual' in this respect: https://github.com/plotly/dash-docs/#running-an-app-locally

File: tests/test_gene_expression_reader.py (hunk 1 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: I hadn't thought of it this way; I'm used to writing a docstring for any function out there but, now that you point it out, I admit it doesn't make much sense in the case of unit tests... Let me silence the "missing docstring" warnings in `pylint` ;)

File: tests/test_gene_expression_reader.py (hunk 2 above)
Comment (shammamah) [reply tests/test_gene_expression_reader.py]: @mkcor Yep -- Python has this package: https://pypi.org/project/GEOparse/

File: tests/test_gene_expression_reader.py (hunk 6 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: Addressed in b04fa7a; the test function name reads "test <given function> <given parameter>" which sounds right to me when testing that it works smoothly, but won't work so well when testing bad format and edge/boundary cases.

> another tip for the unitary test function name, if we end up making a very long test name, it's a good sign we should probably split the scope.

I haven't fully wrapped my head around this one yet... It seems to me that splitting the scope wouldn't lead to even longer function names, because then you would add a sort of prefix to specify the scope...

File: tests/test_gene_expression_reader.py (hunk 1 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: Done via 0ff4381 and b04fa7a.

File: README.md (hunk 3 above)
Comment (mkcor) [reply README.md]: > Other Dash repos use `virtualenv`. I think it might be better to keep consistent with that.

Done via c4ccaf9c08319e5.

File: tests/test_gene_expression_reader.py (hunk 2 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: Handled via 1a24a5151feafa.

File: tests/test_gene_expression_reader.py (hunk 5 above)
Comment (mkcor) [reply tests/test_gene_expression_reader.py]: When testing for the actual values (only returned when `return_filtered_data=True`), I'm running into the exact issue I suspected when I went over the code for:
https://github.com/plotly/dash-bio-utils/blob/b58a2a416bfa51c95328b31ee6c0ab8a7e47f3ce/dash_bio_utils/gene_expression_reader.py#L189-L195
When filtering, `read_soft()` ends up calling this `_get_selected_data()` function, which takes a dataframe as input but spits out an array (hence losing indices and labels, also known as row labels and column labels). As far as `test_read_soft_return_filtered()` is concerned,
https://github.com/plotly/dash-bio-utils/blob/b58a2a416bfa51c95328b31ee6c0ab8a7e47f3ce/dash_bio_utils/gene_expression_reader.py#L209
transforms our labeled data
```py
ipdb> selected_data                                                                                               
           GSM1110880  GSM1110881
ID_REF                           
1053_at       25.7486     25.9348
1007_s_at     25.4542     25.6869
117_at        21.4459     22.1360
ipdb> type(selected_data)                                                                                         
<class 'pandas.core.frame.DataFrame'>
```
into
```py
ipdb> selected_data.values                                                                                        
array([[25.7486, 25.9348],
       [25.4542, 25.6869],
       [21.4459, 22.136 ]])
ipdb> type(selected_data.values)                                                                                  
<class 'numpy.ndarray'>
```
I would then like to test the expected values against
```py
ipdb> import numpy as np
ipdb> np.array([[25.7486, 25.9348], 
       [25.4542, 25.6869], 
       [21.4459, 22.136 ]]) 
```
but row order isn't preserved, which I could handle with by merging and dropping duplicates, but neither is column order (which I cannot handle in such a deterministic test). Otherwise I could test for the 'presence' of all expected values, but that seems weird and weak...

I would be in favour of keeping the data labelled in the parsers, and dropping the labels (i.e., turning the dataframe into an array) only when necessary in the components or apps... @shammamah What do you think?

File: README.md [hunk 7]
```diff
@@ -17,4 +17,39 @@ and more, please [reach out](https://plotly.typeform.com/to/mH1Cpb).
 
 If you would like to contribute to this repository, please refer to
 the [contributing
-guidelines](https://github.com/plotly/dash-bio-utils/blob/master/CONTRIBUTING.md).
\ No newline at end of file
+guidelines](https://github.com/plotly/dash-bio-utils/blob/master/CONTRIBUTING.md).
```
[the remark this answers is not in the document]
Comment (mkcor) [reply README.md]: @shammamah because you are `emacs` and I'm `vim`!?! :smile: 
I think we've discussed it before: We want newlines at EOF (if you check out the plain text diff, you will see "\ No newline at end of file"). All I know is that, with default settings, `vim` includes a newline at EOF...
